### DISCUSSION TITLE
Is `check_same_thread=False` Safe?

### DISCUSSION THREAD

**Discussion by mirusu400** [fastapi/fastapi#5199] [Questions] [answered]
On FastAPI docs, https://fastapi.tiangolo.com/tutorial/sql-databases/, We use `check_same_thread = False` in docs source code.
```python
from sqlalchemy import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker

SQLALCHEMY_DATABASE_URL = "sqlite:///./sql_app.db"
# SQLALCHEMY_DATABASE_URL = "postgresql://user:password@postgresserver/db"

engine = create_engine(
    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}
)
SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

Base = declarative_base()
```

But on sqlite3 documentation,

---

check_same_thread ([bool](https://docs.python.org/ko/3/library/functions.html#bool)) – If [True](https://docs.python.org/ko/3/library/constants.html#True) (default), only the creating thread may use the connection. If [False](https://docs.python.org/ko/3/library/constants.html#False), the connection may be shared across multiple threads; **if so, write operations should be serialized by the user to avoid data corruption.**

---
 I know that `FastAPI` and `uvicorn` works concurrently, but not multithreaded / multiprocessed. but in [gunicorn](https://fastapi.tiangolo.com/deployment/server-workers/) supports multiple workers(procees), it may serve multiple transactions / querys at same time. 

In this situation, is FastAPI guarantee documented methods are safe?

**Comment by JarroVGIT** [comment, marked as answer]
It's important to note that sharing of a connection across threads is very different than sharing it across processes. If you run Gunicorn (or Uvicorn, both have this capability) with multiple workers, all these workers have their own python **process**. That means, they will all create their own connection and it will not be shared. Within the context of one worker, there is (edge cases not considered) only 1 instance of your app, but many requests may be handled by different threads **within that worker context**. 

So, FastAPI actually does work multithreaded (within the limitation of Python) and **not** concurrently, nor multiprocessed. 

In the case of SQLite, it is not really a problem and even required. Some requests are handled in thread 1, while parts of the handling takes place in thread 2. Being able to share the connection across threads will allow for this model. I am not sure about any guarantees though!
